# Data Preparation and Cleaning
In this notebook I'll clean the dataset obtained from kaggle and save it so I can perform analysis.

## Loading libraries

In [343]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 
import seaborn as sns 

## Loading the dataset
### Lead interaction Details 

In [344]:
# Load the interaction details 
interactions_df = pd.read_csv('leads_interaction_details.csv')
interactions_df.head(7)

jnr_sm_id  lead_id     lead_stage call_done_date call_status  \
0  JNR1001MG  USR1001           lead       1/2/2022  successful   
1  JNR1001MG  USR1001           lead       1/2/2022  successful   
2  JNR1001MG  USR1002           lead       1/3/2022  successful   
3  JNR1001MG  USR1002           lead       1/4/2022  successful   
4  JNR1001MG  USR1002      awareness       1/5/2022  successful   
5  JNR1001MG  USR1002      awareness       1/6/2022  successful   
6  JNR1001MG  USR1002  consideration       1/7/2022  successful   

                  call_reason  
0           lead_introduction  
1               demo_schedule  
2           lead_introduction  
3               demo_schedule  
4          post_demo_followup  
5  followup_for_consideration  
6   interested_for_conversion

In [345]:
interactions_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2192 entries, 0 to 2191
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   jnr_sm_id       2192 non-null   object
 1   lead_id         2192 non-null   object
 2   lead_stage      2192 non-null   object
 3   call_done_date  2192 non-null   object
 4   call_status     2192 non-null   object
 5   call_reason     2192 non-null   object
dtypes: object(6)
memory usage: 102.9+ KB


The dataset has 2192 rows and 5 columns. There are no missing values.

In [346]:
interactions_df.duplicated().any()

True

The dataset has duplicated rows. 

In [347]:
interactions_df.describe()

jnr_sm_id  lead_id lead_stage call_done_date call_status  \
count        2192     2192       2192           2192        2192   
unique         16      358          4             40           2   
top     JNR1012MG  USR1193       lead       1/8/2022  successful   
freq          164       11       1547            156        1866   

              call_reason  
count                2192  
unique                  8  
top     demo_not_attended  
freq                  701

### Client basic details

In [348]:
# Load the client details
leads_details = pd.read_csv('leads_basic_details.csv')
leads_details.sample(10)

lead_id  age  gender   current_city current_education  \
103  USR1104   16    MALE         Mumbai      Intermediate   
123  USR1124   22  FEMALE          Kochi   Looking for Job   
52   USR1053   24    MALE  Visakhapatnam   Looking for Job   
33   USR1034   16    MALE      Bengaluru      Intermediate   
129  USR1130   21  FEMALE          Kochi            Degree   
16   USR1017   24    MALE      Hyderabad   Looking for Job   
56   USR1057   22    MALE        Chennai   Looking for Job   
24   USR1025   25  FEMALE          Kochi   Looking for Job   
203  USR1204   16    MALE         Mumbai      Intermediate   
157  USR1158   25    MALE        Chennai            B.Tech   

       parent_occupation  lead_gen_source  
103          IT Employee   user_referrals  
123          IT Employee     social_media  
52              Business              SEO  
33   Government Employee     social_media  
129  Government Employee  email_marketing  
16                Lawyer     social_media  
56   Government Employee          website  
24      Private Employee  email_marketing  
203          IT Employee   user_referrals  
157  Government Employee              SEO

In [349]:
leads_details.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 360 entries, 0 to 359
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   lead_id            360 non-null    object
 1   age                360 non-null    int64 
 2   gender             360 non-null    object
 3   current_city       360 non-null    object
 4   current_education  360 non-null    object
 5   parent_occupation  360 non-null    object
 6   lead_gen_source    360 non-null    object
dtypes: int64(1), object(6)
memory usage: 19.8+ KB


The datset contains details of 360 clients with information like age, gender, current city and education, parent's occupation and how the client came to know about the product.

In [350]:
leads_details.duplicated().any()

False

There are no duplicated rows. 

In [351]:
leads_details.describe()

age
count  360.000000
mean    21.561111
std     11.555444
min     16.000000
25%     18.000000
50%     21.000000
75%     24.000000
max    211.000000

There are outliers in this dataset.

In [352]:
demos_df = pd.read_csv('leads_demo_watched_details.csv')
demos_df.sample(10)

lead_id demo_watched_date language  watched_percentage
25   USR1027          1/9/2022  English                  11
140  USR1227          1/9/2022  English                  73
3    USR1005          1/3/2022    Hindi                  38
131  USR1211         1/26/2022  English                  72
92   USR1135         1/17/2022   Telugu                  65
57   USR1070         1/11/2022  English                  71
176  USR1295         1/18/2022  English                  24
26   USR1028         1/10/2022   Telugu                  50
100  USR1151         1/18/2022   Telugu                  37
19   USR1021          1/2/2022  English                 100

In [353]:
demos_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 194 entries, 0 to 193
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   lead_id             194 non-null    object
 1   demo_watched_date   194 non-null    object
 2   language            194 non-null    object
 3   watched_percentage  194 non-null    int64 
dtypes: int64(1), object(3)
memory usage: 6.2+ KB


The dataset contains details of demos watched by leads. There are no missing values. 

In [354]:
demos_df.describe()

watched_percentage
count          194.000000
mean            56.634021
std             43.555635
min              2.000000
25%             35.000000
50%             55.500000
75%             75.750000
max            510.000000

There are leads who watched 510% of the demo. This is ofcourse is an outlier so it should be cleaned.

### Reasons for not being interested
This dataset contains the leads' reasons as to why they dropped out at the various stages of the customer acquistion process.

In [355]:
no_interest_df = pd.read_csv('leads_reasons_for_no_interest.csv')
no_interest_df

lead_id reasons_for_not_interested_in_demo  \
0    USR1001                No time for student   
1    USR1003                                NaN   
2    USR1004                                NaN   
3    USR1005                                NaN   
4    USR1006                                NaN   
..       ...                                ...   
289  USR1356                      Cannot afford   
290  USR1357                      Cannot afford   
291  USR1358              Wants offline classes   
292  USR1359            Will join in final year   
293  USR1360            Will join in final year   

    reasons_for_not_interested_to_consider  \
0                                      NaN   
1                      No time for student   
2                    Wants offline classes   
3                             Can't afford   
4         Student not interested in domain   
..                                     ...   
289                                    NaN   
290                                    NaN   
291                                    NaN   
292                                    NaN   
293                                    NaN   

    reasons_for_not_interested_to_convert  
0                                     NaN  
1                                     NaN  
2                                     NaN  
3                                     NaN  
4                                     NaN  
..                                    ...  
289                                   NaN  
290                                   NaN  
291                                   NaN  
292                                   NaN  
293                                   NaN  

[294 rows x 4 columns]

In [356]:
no_interest_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294 entries, 0 to 293
Data columns (total 4 columns):
 #   Column                                  Non-Null Count  Dtype 
---  ------                                  --------------  ----- 
 0   lead_id                                 294 non-null    object
 1   reasons_for_not_interested_in_demo      164 non-null    object
 2   reasons_for_not_interested_to_consider  79 non-null     object
 3   reasons_for_not_interested_to_convert   51 non-null     object
dtypes: object(4)
memory usage: 9.3+ KB


We have missing values in the last 3 columns. We can melt them to 2 columns, which stage they dropped out and the reason.

In [357]:
for col in no_interest_df.columns[1:]:
    print(no_interest_df[col].unique())

['No time for student' nan 'Student not interested in domain'
 'Wants offline classes' "Can't afford" 'Cannot afford'
 'Will join in final year']
[nan 'No time for student' 'Wants offline classes' "Can't afford"
 'Student not interested in domain' 'Will join in final year']
[nan "Can't afford" 'Will join in final year' 'No time for student'
 'Student not interested in domain' 'Wants offline classes']


Reasons like `Can't afford` and `Cannot afford` are the same, so it can be combined into one.

## Cleaning the data and wrangling it

In [358]:
# Making copies of the dataframe
leads_details_clean = leads_details.copy()
interactions_clean = interactions_df.copy()
demos_clean = demos_df.copy()
no_interest_clean = no_interest_df.copy()

### Remove the duplicated rows
**Task: Remove duplicate rows**

**Code:**

In [359]:
interactions_clean = interactions_clean.drop_duplicates()

**Test:**

In [360]:
interactions_clean.duplicated().any()

False

### Remove outliers
**Task: Remove outliers in the lead details**

**Code:**

In [361]:
sorted(leads_details.age.unique())

[16, 18, 20, 21, 22, 24, 25, 116, 211]

Any age greater than 25 is an outlier

In [362]:
leads_details_clean = leads_details_clean.query("age <= 25")

**Test:**

In [363]:
leads_details_clean.age.describe()

count    358.000000
mean      20.768156
std        2.852959
min       16.000000
25%       18.000000
50%       21.000000
75%       24.000000
max       25.000000
Name: age, dtype: float64

**Task: Remove outliers in the demo dataset**

**Code:**

In [364]:
## Anything greater than 100% is an outlier
demos_clean = demos_clean.query('watched_percentage<=100')

**Test:**

In [365]:
demos_clean.watched_percentage.max()

100

**Task: Combining reasons columns into 2 columns, stage dropped out and the reason**

**Code:**

In [366]:
stages = {'reasons_for_not_interested_in_demo':'lead',
          'reasons_for_not_interested_to_consider':'awareness',
          'reasons_for_not_interested_to_convert':'consider'}
# Melts the reasons columns into 2 columns
no_interest_clean = no_interest_clean.melt('lead_id',no_interest_clean.columns[1:4],'stage_dropped','reason').dropna()
# Makes sure the reasons are unique
no_interest_clean.loc[no_interest_clean.reason == 'Cannot afford',"reason"] = "Can't afford"
no_interest_clean.loc[no_interest_clean.reason == "Student not interested in domain","reason"] = "Not Interested in domain"
# Replaces the melted column names with the actual stages
no_interest_clean['stage_dropped'] = no_interest_clean['stage_dropped'].apply(lambda x: stages[x])

In [367]:
no_interest_clean[no_interest_clean['lead_id']=='USR1015']

lead_id stage_dropped                 reason
303  USR1015     awareness  Wants offline classes

**Test:**

In [368]:
no_interest_clean

lead_id stage_dropped                    reason
0    USR1001          lead       No time for student
32   USR1048          lead  Not Interested in domain
33   USR1049          lead       No time for student
35   USR1054          lead     Wants offline classes
37   USR1056          lead              Can't afford
..       ...           ...                       ...
819  USR1287      consider     Wants offline classes
823  USR1291      consider     Wants offline classes
832  USR1305      consider     Wants offline classes
864  USR1343      consider     Wants offline classes
869  USR1348      consider     Wants offline classes

[294 rows x 3 columns]

The dataset has now 3 columns, lead_id, stage_dropped and reason

In [369]:
no_interest_clean['reason'].unique()

array(['No time for student', 'Not Interested in domain',
       'Wants offline classes', "Can't afford", 'Will join in final year'],
      dtype=object)

Unique responses

## Saving the cleaned data

In [370]:
leads_details_clean.to_csv('clean/leads_basic_details_clean.csv',index=False)
no_interest_clean.to_csv('clean/leads_reasons_for_no_interest_clean.csv',index=False)
interactions_clean.to_csv('clean/leads_interaction_details_clean.csv',index=False)
demos_clean.to_csv('clean/leads_demo_watched_details_clean.csv',index=False)